pkg/crypto/certloader/config.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package certloader
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io/ioutil"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/sirupsen/logrus"
+
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+// WatchedConfig is a set of TLS configuration files including CA files, and a
+// certificate along with its private key. The files are watched for change and
+// reloaded automatically.
+type WatchedConfig struct {
+	log         logrus.FieldLogger
+	caFiles     []string
+	certFile    string
+	privkeyFile string
+	stop        chan struct{}
+	mu          lock.Mutex
+	caCertPool  *x509.CertPool
+	keypair     *tls.Certificate
+}
+
+var (
+	// ErrCertWithoutPrivkey is returned when a certificate is configured
+	// without its corresponding private key.
+	ErrCertWithoutPrivkey = errors.New("certificate and private key are both required, but only the certificate was provided")
+	// ErrPrivkeyWithoutCert is returned when a private key is configured
+	// without its corresponding certificate.
+	ErrPrivkeyWithoutCert = errors.New("certificate and private key are both required, but only the private key was provided")
```

Comment:
```suggestion
	// ErrInvalidKeypair is returned when one of the certificate or corresponding private key of a keypair is missing.
	ErrInvalidKeypair = errors.New("certificate and private key are both required, but only one was provided")
```

I think this could be a single error as failing to provide both the cert and corresponding private key is essentially a single error. This would allow the caller to only check for a single error when need be. One could argue that telling which of the cert or key is missing provides additional information but I think that in a real scenario, figuring out which one is missing should be easy.